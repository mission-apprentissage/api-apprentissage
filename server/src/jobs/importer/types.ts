import type { CronDef } from "job-processor";
import type { ImportStatus } from "shared";

export type Importer = CronDef & {
  getStatus(): Promise<ImportStatus>;
};
