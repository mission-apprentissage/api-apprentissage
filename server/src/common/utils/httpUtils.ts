import axios, { AxiosInstance, AxiosRequestConfig, CreateAxiosDefaults } from "axios";
import http from "http";
import https from "https";
import iconv from "iconv-lite";
import { compose } from "oleoduc";

import logger from "@/common/logger";

// https://github.com/axios/axios/issues/3845#issuecomment-1040819908
class BufferedHttpAgent extends http.Agent {
  highWaterMark: number;

  constructor({ highWaterMark = 16 * 1024, ...rest }) {
    super(rest);

    this.highWaterMark = highWaterMark;
  }
  // @ts-expect-error: TODO
  createConnection(options, callback) {
    // @ts-expect-error
    return super.createConnection({ ...options, highWaterMark: this.highWaterMark }, callback);
  }
}
class BufferedHttpsAgent extends https.Agent {
  highWaterMark: number;

  constructor({ highWaterMark = 16 * 1024, ...rest }) {
    super(rest);
    //see https://github.com/nodejs/node/issues/39092
    this.highWaterMark = highWaterMark;
  }

  // @ts-expect-error: TODO
  createConnection(options, callback) {
    // @ts-expect-error
    return super.createConnection({ ...options, highWaterMark: this.highWaterMark }, callback);
  }
}

type FetchOptions = AxiosRequestConfig & {
  highWaterMark: number;
  raw: boolean;
  encoding: string;
};
async function _fetch(url: string, options: Partial<FetchOptions> = {}, client: AxiosInstance = axios) {
  const { method = "GET", data, highWaterMark, ...rest } = options;
  logger.debug(`${method} ${url}...`);

  return client.request({
    url,
    method,
    httpAgent: new BufferedHttpAgent({ highWaterMark }),
    httpsAgent: new BufferedHttpsAgent({ highWaterMark }),
    ...(data ? { data } : {}),
    ...rest,
  });
}

async function fetchStream(url: string, options: Partial<FetchOptions> = {}, client: AxiosInstance = axios) {
  const { raw, encoding = "UTF-8", ...rest } = options;
  const response = await _fetch(url, { responseType: "stream", ...rest }, client);

  const stream = compose(response.data, iconv.decodeStream(encoding));

  if (raw) {
    return {
      headers: response.headers,
      stream,
    };
  } else {
    return stream;
  }
}

async function fetchData(url: string, options: Partial<FetchOptions> = {}, client: AxiosInstance = axios) {
  const { raw, ...rest } = options;
  const response = await _fetch(url, { responseType: "json", ...rest }, client);
  const data = response.data;

  if (raw) {
    return {
      headers: response.headers,
      data,
    };
  } else {
    return data;
  }
}

async function fetchJson(url: string, options: Partial<FetchOptions> = {}) {
  const response = await _fetch(url, { ...options, responseType: "json" });
  return response.data;
}

function addCsvHeaders(filename: string, encoding: string, res: any) {
  res.setHeader("Content-disposition", `attachment; filename=${filename}`);
  res.setHeader("Content-Type", `text/csv; charset=${encoding}`);
}

const getHttpClient = (options: CreateAxiosDefaults<any> = {}) =>
  axios.create({
    timeout: 15000,
    httpAgent: new http.Agent({ keepAlive: true }),
    httpsAgent: new https.Agent({ keepAlive: true }),
    ...options,
  });

export { addCsvHeaders, fetchJson, fetchStream, fetchData, getHttpClient };
