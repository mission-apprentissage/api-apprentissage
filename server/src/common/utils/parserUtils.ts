import { Options, parse } from "csv-parse";
import { isEmpty, pickBy } from "lodash-es";

export const DEFAULT_DELIMITER = ";";

export function parseCsv(options: Options = {}) {
  return parse({
    trim: true,
    delimiter: DEFAULT_DELIMITER,
    columns: true,
    bom: true,
    relax_column_count: true,
    on_record: (record) => {
      return pickBy(record, (v) => {
        return !isEmpty(v) && v.trim().length;
      });
    },
    ...options,
  });
}
