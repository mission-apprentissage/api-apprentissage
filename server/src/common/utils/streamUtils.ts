import { Transform } from "node:stream";

type Options = {
  size: number;
};

export function createBatchTransformStream(opts: Options): Transform {
  let currentBatch: unknown[] = [];

  return new Transform({
    objectMode: true,
    transform(chunk, _encoding, callback) {
      currentBatch.push(chunk);

      if (currentBatch.length >= opts.size) {
        this.push(currentBatch);
        currentBatch = [];
      }

      callback();
    },
    flush(callback) {
      if (currentBatch.length > 0) {
        this.push(currentBatch);
      }
      callback();
    },
  });
}
