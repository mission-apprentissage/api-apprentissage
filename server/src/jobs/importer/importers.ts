import { getAcceeImporterStatus, runAcceImporter } from "./acce/acce.js";
import { getBcnImporterStatus, runBcnImporter } from "./bcn/bcn.importer.js";
import { getCatalogueImporterStatus, runCatalogueImporter } from "./catalogue/catalogue.importer.js";
import { getCertificationImporterStatus, importCertifications } from "./certifications/certifications.importer.js";
import { getCommuneImporterStatus, runCommuneImporter } from "./commune/commune.importer.js";
import { getDaresApiIdccImporterStatus, runDaresApeIdccImporter } from "./dares/ape_idcc/dares.ape_idcc.importer.js";
import { getDaresCcnImporterStatus, runDaresConventionCollectivesImporter } from "./dares/ccn/dares.ccn.importer.js";
import { getFormationsImporterStatus, importFormations } from "./formation/formation.importer.js";
import { getFranceCompetencesImporterStatus, runRncpImporter } from "./france_competence/france_competence.importer.js";
import { getKaliImporterStatus, runKaliConventionCollectivesImporter } from "./kali/kali.ccn.importer.js";
import { getKitApprentissageImporterStatus, runKitApprentissageImporter } from "./kit/kitApprentissage.importer.js";
import { getMissionLocaleImporterStatus, runMissionLocaleImporter } from "./mission_locale/mission_locale.importer.js";
import { getOrganismesImporterStatus, importOrganismes } from "./organisme/organisme.importer.js";
import { getReferentielImporterStatus, runReferentielImporter } from "./referentiel/referentiel.js";
import type { Importer } from "./types.js";

const timings = {
  import_source: "0 4 * * *",
  import_source_main: "0 */4 * * *",
};

export const importers: Record<string, Importer> = {
  "Mise à jour acce": {
    cron_string: timings.import_source,
    handler: runAcceImporter,
    resumable: true,
    getStatus: getAcceeImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des données BCN": {
    cron_string: timings.import_source_main,
    handler: runBcnImporter,
    resumable: true,
    getStatus: getBcnImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des données Kit Apprentissage": {
    cron_string: timings.import_source_main,
    handler: runKitApprentissageImporter,
    resumable: true,
    getStatus: getKitApprentissageImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des données Referentiel": {
    cron_string: timings.import_source_main,
    handler: runReferentielImporter,
    resumable: true,
    getStatus: getReferentielImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des données Catalogue": {
    cron_string: timings.import_source_main,
    handler: runCatalogueImporter,
    resumable: true,
    getStatus: getCatalogueImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des données France Compétences": {
    cron_string: timings.import_source,
    handler: runRncpImporter,
    resumable: true,
    getStatus: getFranceCompetencesImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des certifications": {
    cron_string: timings.import_source_main,
    handler: async () => importCertifications(),
    resumable: true,
    getStatus: getCertificationImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des organismes": {
    cron_string: timings.import_source_main,
    handler: async () => importOrganismes(false),
    resumable: true,
    getStatus: getOrganismesImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des formations": {
    cron_string: timings.import_source_main,
    handler: importFormations,
    resumable: true,
    getStatus: getFormationsImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des Conventions Collective Kali": {
    cron_string: timings.import_source,
    handler: runKaliConventionCollectivesImporter,
    resumable: true,
    getStatus: getKaliImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des Conventions Collective Dares": {
    cron_string: timings.import_source,
    handler: runDaresConventionCollectivesImporter,
    resumable: true,
    getStatus: getDaresCcnImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des APE-IDCC Dares": {
    cron_string: timings.import_source,
    handler: runDaresApeIdccImporter,
    resumable: true,
    getStatus: getDaresApiIdccImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des Missions Locales": {
    cron_string: timings.import_source,
    handler: runMissionLocaleImporter,
    resumable: true,
    getStatus: getMissionLocaleImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des Communes": {
    cron_string: timings.import_source,
    handler: runCommuneImporter,
    resumable: true,
    getStatus: getCommuneImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
};
