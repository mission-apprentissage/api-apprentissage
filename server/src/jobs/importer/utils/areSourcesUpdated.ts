export function areSourcesUpdated(
  lastImportSources: Record<string, { import_date: Date }> | null | undefined,
  currentSources: Record<string, { import_date: Date }>
): boolean {
  if (!lastImportSources) {
    return true;
  }

  for (const source in currentSources) {
    if (!lastImportSources[source] || lastImportSources[source].import_date < currentSources[source].import_date) {
      return true;
    }
  }

  return false;
}
