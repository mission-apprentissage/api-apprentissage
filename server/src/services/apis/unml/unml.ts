import { internal, isBoom } from "@hapi/boom";
import { isAxiosError } from "axios";
import type { ISourceUnmlPayload } from "shared";
import { zSourceUnmlPayload } from "shared";

import config from "@/config.js";
import getApiClient from "@/services/apis/client.js";
import { withCause } from "@/services/errors/withCause.js";
import { apiRateLimiter } from "@/utils/apiUtils.js";

const unmlClient = apiRateLimiter("unml", {
  nbRequests: 10,
  durationInSeconds: 1,
  client: getApiClient(
    {
      baseURL: config.api.unml.endpoint,
    },
    { cache: false }
  ),
});

export async function fetchDepartementMissionLocale(codeDepartement: string): Promise<ISourceUnmlPayload> {
  return unmlClient(async (client) => {
    try {
      const { data } = await client.get(`/TrouveTaML/?search_=${codeDepartement}`);

      const result = zSourceUnmlPayload.safeParse(data);

      if (!result.success) {
        throw internal("api.unml: unable to parse payload", { data, error: result.error });
      }

      return result.data;
    } catch (error) {
      if (isBoom(error)) {
        throw error;
      }

      if (isAxiosError(error)) {
        throw internal("api.unml: unable to fetchDepartementMissionLocale", { data: error.toJSON() });
      }
      throw withCause(internal("api.unml: unable to fetchDepartementMissionLocale"), error);
    }
  });
}
