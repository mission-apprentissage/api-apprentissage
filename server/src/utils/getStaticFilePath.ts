import path from "path";
import { __dirname } from "api-alternance-sdk/internal";

export function getStaticDirPath(): string {
  // tsup build project into a single file to dist/index.js
  // Hence import.meta.url references dist/index.js
  if (process.env.IS_BUILT === "true") {
    return path.join(__dirname(import.meta), "../static");
  }

  // When ran directly from source code with tsx or jest
  // Then files are not compiled into a single file
  return path.join(__dirname(import.meta), "../static");
}

export function getStaticFilePath(relativeFilename: string): string {
  return path.join(getStaticDirPath(), relativeFilename);
}
