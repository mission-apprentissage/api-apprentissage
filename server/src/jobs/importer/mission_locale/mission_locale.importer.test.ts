import { useMongo } from "@tests/mongo.test.utils.js";
import { ObjectId } from "mongodb";
import { dirname, join } from "path";
import { sourceUnmlResultsFixtures } from "shared/models/fixtures/commune.model.fixture";
import { fileURLToPath } from "url";
import { beforeEach, describe, expect, it, vi } from "vitest";

import { fetchDepartementMissionLocale } from "@/services/apis/unml/unml.js";
import { getDbCollection } from "@/services/mongodb/mongodbService.js";
import { getStaticFilePath } from "@/utils/getStaticFilePath.js";

import { runMissionLocaleImporter } from "./mission_locale.importer.js";

useMongo();

vi.mock("@/utils/getStaticFilePath", () => ({
  getStaticFilePath: vi.fn(),
}));
vi.mock("@/services/apis/unml/unml.js");

describe("runMissionLocaleImporter", () => {
  const now = new Date("2024-10-03T21:53:08.141Z");

  beforeEach(() => {
    vi.useFakeTimers();
    vi.setSystemTime(now);

    return () => vi.useRealTimers();
  });

  it("should import initial ml", async () => {
    vi.mocked(getStaticFilePath).mockImplementation(() =>
      join(dirname(fileURLToPath(import.meta.url)), `fixtures/couverture.csv`)
    );
    vi.mocked(fetchDepartementMissionLocale).mockImplementation(
      async (codeDepartement: string) =>
        sourceUnmlResultsFixtures[codeDepartement as keyof typeof sourceUnmlResultsFixtures]
    );

    await runMissionLocaleImporter();
    expect(fetchDepartementMissionLocale).toHaveBeenCalledTimes(4);

    const sources = await getDbCollection("source.insee_to_ml").find({}).toArray();

    expect(
      sources.map(({ _id, import_id, ...rest }) => rest).toSorted((a, b) => a.code_insee.localeCompare(b.code_insee))
    ).toMatchSnapshot();

    const importMetas = await getDbCollection("import.meta").find().toArray();

    expect(importMetas).toEqual([
      {
        _id: expect.any(ObjectId),
        import_date: now,
        type: "mission_locale",
        status: "done",
      },
    ]);
  });
});
