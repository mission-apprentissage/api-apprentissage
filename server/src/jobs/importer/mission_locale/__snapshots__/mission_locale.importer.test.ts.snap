// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runMissionLocaleImporter > should import initial ml 1`] = `
[
  {
    "code_insee": "18288",
    "ml": null,
  },
  {
    "code_insee": "18289",
    "ml": null,
  },
  {
    "code_insee": "75101",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75102",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75103",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75104",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75105",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75106",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75107",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75108",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75109",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75110",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75111",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75112",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75113",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75114",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75115",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75116",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75117",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75118",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75119",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "75120",
    "ml": {
      "code": "75119",
      "contact": {
        "email": "contact@mail.fr",
        "siteWeb": "https://www.missionlocale.paris/",
        "telephone": "0179970000",
      },
      "id": 226,
      "localisation": {
        "adresse": "22 rue Pajol ",
        "cp": "75018",
        "geopoint": {
          "coordinates": [
            2.3740736,
            48.8848179,
          ],
          "type": "Point",
        },
        "ville": "PARIS",
      },
      "nom": "DE PARIS",
      "siret": "53132862300149",
    },
  },
  {
    "code_insee": "77001",
    "ml": {
      "code": "77333",
      "contact": {
        "email": null,
        "siteWeb": "https://www.mlseineetloing.fr/",
        "telephone": "01 64 28 51 58",
      },
      "id": 374,
      "localisation": {
        "adresse": "17 Rue des Tanneurs ",
        "cp": "77140",
        "geopoint": {
          "coordinates": [
            2.6988983,
            48.2655271,
          ],
          "type": "Point",
        },
        "ville": "NEMOURS",
      },
      "nom": "DE LA SEINE ET DU LOING",
      "siret": "44270199100014",
    },
  },
  {
    "code_insee": "77002",
    "ml": null,
  },
  {
    "code_insee": "77003",
    "ml": {
      "code": "77333",
      "contact": {
        "email": null,
        "siteWeb": "https://www.mlseineetloing.fr/",
        "telephone": "01 64 28 51 58",
      },
      "id": 374,
      "localisation": {
        "adresse": "17 Rue des Tanneurs ",
        "cp": "77140",
        "geopoint": {
          "coordinates": [
            2.6988983,
            48.2655271,
          ],
          "type": "Point",
        },
        "ville": "NEMOURS",
      },
      "nom": "DE LA SEINE ET DU LOING",
      "siret": "44270199100014",
    },
  },
]
`;
